src/components/device/ha-device-trigger-picker.ts

```diff
@@ -0,0 +1,184 @@
+import "@polymer/paper-input/paper-input";
+import "@polymer/paper-item/paper-item";
+import "@polymer/paper-item/paper-item-body";
+import "@polymer/paper-dropdown-menu/paper-dropdown-menu-light";
+import "@polymer/paper-listbox/paper-listbox";
+import {
+  LitElement,
+  TemplateResult,
+  html,
+  css,
+  CSSResult,
+  customElement,
+  property,
+} from "lit-element";
+import { HomeAssistant } from "../../types";
+import { fireEvent } from "../../common/dom/fire_event";
+import computeStateName from "../../common/entity/compute_state_name";
+import {
+  DeviceTrigger,
+  fetchDeviceTriggers,
+  triggersEqual,
+} from "../../data/device_automation";
+
+@customElement("ha-device-trigger-picker")
+class HaDeviceTriggerPicker extends LitElement {
+  public hass?: HomeAssistant;
+  @property() public label?: string;
+  @property() public deviceId?: string;
+  @property() public triggers: { [key: string]: DeviceTrigger } = {};
+  public presetTrigger?: DeviceTrigger;
+
+  private noTrigger: DeviceTrigger = {
+    device_id: "",
+    platform: "device",
+    domain: "",
+    entity_id: "",
+  };
+  private unknownTrigger?: DeviceTrigger = {
+    device_id: "",
+    platform: "device",
+    domain: "",
+    entity_id: "",
+  };
+  private key?: string;
+  private setTrigger?: DeviceTrigger;
+
+  private readonly NO_TRIGGER = "NO_TRIGGER";
+  private readonly UNKNOWN_TRIGGER = "UNKNOWN_TRIGGER";
+
+  protected render(): TemplateResult | void {
+    const noTriggers = Object.keys(this.triggers).length === 0;
+    return html`
+      <paper-dropdown-menu-light .label=${this.label} ?disabled=${noTriggers}>
+        <paper-listbox
+          slot="dropdown-content"
+          .selected=${this._key}
+          attr-for-selected="key"
+          @iron-select=${this._triggerChanged}
+          id="listbox"
+        >
+          <paper-item .key=${this.NO_TRIGGER} .trigger=${this.noTrigger} hidden>
+            No triggers
+          </paper-item>
+          <paper-item
+            .key=${this.UNKNOWN_TRIGGER}
+            .trigger=${this.unknownTrigger}
+            hidden
+          >
+            Unknown trigger
+          </paper-item>
+          ${Object.keys(this.triggers).map(
```

Comment:
Reading from the code, creating an object to maintain triggers is causing more complicated code. There is only one place you do a lookup, so I would keep it as an array. That way we are also guaranteed about the order.

Reply:
```ts
this.triggers.map(trigger => html``)
```

Reply:
Yes, but the problem is the cases where there is no trigger, or where the trigger in the automation is not matching the trigger listed by the device.

Reply:
but you can still render those above the list, like you do here ?